### PR TITLE
Migrate ERC20

### PR DESCRIPTION
**Pull request by martriay**
Continuation of https://github.com/OpenZeppelin/cairo-contracts/pull/575 started by @andrew-fleming

### CODE REVIEW

File: src/openzeppelin/token/erc20.cairo [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+            let caller = get_caller_address();
+            _approve(caller, spender, amount);
+            true
+        }
+    }
+
+    #[constructor]
+    fn constructor(name: felt, symbol: felt, initial_supply: u256, recipient: ContractAddress) {
+        initializer(name, symbol, initial_supply, recipient);
+    }
```
Comment (Amxx) [inline src/openzeppelin/token/erc20.cairo:83]: In understand that we may want the constructor to mint an initial supply, but I'm not sure that should be part of the `initializer`

```suggestion
    fn constructor(name: felt, symbol: felt, initial_supply: u256, recipient: ContractAddress) {
        initializer(name, symbol);
        _mint(recipient, initial_supply);
    }
```

File: src/openzeppelin/token/erc20.cairo [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+
+    ///
+    /// Internals
+    ///
+
+    fn initializer(name_: felt, symbol_: felt, initial_supply: u256, recipient: ContractAddress) {
+        _name::write(name_);
+        _symbol::write(symbol_);
+        _mint(recipient, initial_supply);
+    }
```
Comment (Amxx) [inline src/openzeppelin/token/erc20.cairo:148]: see comment in the constructor
```suggestion
    fn initializer(name_: felt, symbol_: felt) {
        _name::write(name_);
        _symbol::write(symbol_);
    }
```

File: src/openzeppelin/token/erc20.cairo (hunk 1 above)
Comment (andrew-fleming) [reply src/openzeppelin/token/erc20.cairo:83]: Agreed, good call

File: src/openzeppelin/token/erc20.cairo [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+
+    fn _increase_allowance(spender: ContractAddress, added_value: u256) -> bool {
+        let caller = get_caller_address();
+        _approve(caller, spender, _allowances::read((caller, spender)) + added_value);
+        true
+    }
+
+    fn _decrease_allowance(spender: ContractAddress, subtracted_value: u256) -> bool {
+        let caller = get_caller_address();
+        _approve(caller, spender, _allowances::read((caller, spender)) - subtracted_value);
```
[the remark this answers is not in the document]
Comment (andrew-fleming) [reply src/openzeppelin/token/erc20.cairo:158]: Correct. The integer checks are in [integer.cairo](https://github.com/starkware-libs/cairo/blob/main/corelib/src/integer.cairo).

And if you test overflows, it'll panic like this:

`failures:
   openzeppelin::tests::test_erc20::test__transfer_not_enough_balance - panicked with [39879774624085075084607933104993585622903 ('u256_sub Overflow'), ].
`

File: src/openzeppelin/token/erc20.cairo [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+
+    fn _spend_allowance(owner: ContractAddress, spender: ContractAddress, amount: u256) {
+        let current_allowance = _allowances::read((owner, spender));
+        let ONES_MASK = 0xffffffffffffffffffffffffffffffff_u128;
+        let is_unlimited_allowance =
+            current_allowance.low == ONES_MASK & current_allowance.high == ONES_MASK;
+        if !is_unlimited_allowance {
+            _approve(owner, spender, current_allowance - amount);
+        }
+    }
```
Comment (Amxx) [inline src/openzeppelin/token/erc20.cairo:199]: With this proposal https://github.com/starkware-libs/cairo/pull/2563 we could do:

```suggestion
    fn _spend_allowance(owner: ContractAddress, spender: ContractAddress, amount: u256) {
        let current_allowance = _allowances::read((owner, spender));
        if is_unlimited_allowance != max_uint256() {
            _approve(owner, spender, current_allowance - amount);
        }
    }
```

**Comment by martriay** [comment]
Superseded by https://github.com/OpenZeppelin/cairo-contracts/pull/586

**Pull request closed without merging**